README.md
# socklog
Active maintained fork of Gerrit Pape's socklog to compliment the corresponding runit one.


CHANGES.md
socklog 2.2.0 
Wed, 03 Mar 2019 19:30:00 CST
  * Resurrected this to be buildable as a clean cross compile
    just like runit was done a bit under two years ago now.
  * Moved to a CMake based build system for that purpose.
  * Did a similar restructure of the directories to make this
    more maintainable.
  * Added nanoklogd from Void's build out for this package to
    this as it properly belongs in here as far as I'm concerned.

socklog 2.1.0
Mon, 06 Mar 2006 12:59:33 +0000
  * socklog-conf, man/socklog-conf.8: create service directories
    /etc/sv/socklog-* instead of /etc/socklog/*; create ./check script
    for socklog-unix service.
  * doc/configuration.html, doc/install.html, doc/network.html,
    doc/notify.html: adapt to new path to service directories; utilize
    runit's sv program instead of runsvctrl.
  * man/socklog-check.8, man/socklog-conf.8, man/socklog.8, man/uncat.1:
    see also sv(8).

socklog 2.0.3
Sun, 26 Feb 2006 13:29:17 +0000
  * doc/index.html: minor.
  * socklog-check.c: change EDESTADDRREQ to ECONNREFUSED if defined and set
    by connect().
  * tryto: fix a race with child terminating immediatly (thx Alex Efros);
    catch SIGCHLD while waiting ktimeout seconds before sending KILL.
  * man/tryto.1: remove reference to timelimit(8) (thx Alex Efros).
  * socklog.check, socklog.dist: don't check usage output with -V as it
    differs on solaris (thx Asif Iqbal).

socklog 2.0.2
Sun, 10 Apr 2005 07:21:30 +0000
  * doc/readme.solaris.html: add -ldoor -lthread to src/conf-ld (thx Lars
    Kellogg-Stedman); update log/run script, mailing list address.
  * man/socklog.8: remove double entries in SEE ALSO.

socklog 2.0.1
Thu, 10 Feb 2005 10:05:55 +0000
  * Makefile: add linker options detected by trysocketlib.c to targets
    check-socklog-inet, check-socklog-inet (fixes test failure on solaris,
    thx Uffe Jakobsen).

socklog 2.0.0
Sun, 06 Feb 2005 19:57:04 +0000
  * socklog-check.check: tweak timing in program test.
  * tryto.c: new option -P: run prog in new session, and signal process
    group.
  * man/tryto.1: adapt; cleanup.
  * man/socklog-check.8: add hint about runit services using syslog
    facility; cleanup.
  * man/socklog-conf.8, man/socklog.8, man/uncat.1: cleanup.
  * doc/index.html: new stable release.
  * doc/network.html: typo.

socklog 1.5.0
Sat, 26 Jun 2004 11:28:33 +0000
  * socklog-check, man/socklog-check.8: new; check for the availability of
    a socklog service.
  * socklog-check.check, socklog-check.dist: new; program test.
  * byte_zero.c: new.
  * socklog.c, man/socklog.8: new option -U: don't override umask when
    creating domain socket.
  * package/check, package/compile, package/install-man, package/upgrade:
    minor cleanup.
  * Makefile, man/socklog-conf.8: adapt.

socklog 1.4.2
Tue, 23 Mar 2004 18:16:17 +0000
  * doc/index.html, doc/install.html: minor.
  * socklog.c: fix sigterm handling in ucspi mode.
  * uncat.check: tweak timing in program test.
  * Makefile: don't run target check by default.
  * package/check: new; run checks on programs.
  * package/install: run package/check.

socklog 1.4.1
Sat, 28 Feb 2004 17:47:13 +0000
  * socklog.c: handle sigterm on other platforms as accurately as already
    done on solaris.
  * tryto.c, uncat.c: minor cleanup.
  * check-diff, check-dist, check-local: new; run checks on programs.
  * socklog-conf.check, socklog.check, tryto.check, uncat.check,
    check-socklog-inet.c, check-socklog-unix.c: new; check program.
  * socklog-conf.dist, socklog.dist, tryto.dist, uncat.dist: new; dist check
    program output.
  * package/COPYING: 2004.
  * package/compile, package/install-man, package/upgrade: minor.

socklog 1.3.1
Mon, 19 Jan 2004 18:39:01 +0000
  * trysocketlib.c: new; check for libraries needed for socket() on some
    systems (solaris, thx Uffe Jakobsen).
  * Makefile: adapt.
  * socklog.c, socklog-conf.c: add defines automatically detecting solaris
    systems; now builds on solaris out-of-the-box (thx Uffe Jakobsen, Lars
    Uffmann).
  * doc/readme.solaris.html: adapt (thx Lars Uffmann).
  * print-cc.sh, print-ld.sh: head -1 -> head -n1.

socklog 1.3.0
Mon, 12 Jan 2004 19:06:56 +0000
  * socklog.c: add getopt-style command line options -r, -R for logging raw
    syslog messages without priority, facility conversion.
  * man/socklog.8: adapt.
  * doc/network.html: document old-style syslog network logging through UDP.
  * doc/index.html: change short description.
  * package/compile, package/install-man, package/upgrade: exit 1 on
    sub-shell failures.

socklog 1.2.0
Sat, 09 Aug 2003 18:08:20 +0200
  * socklog-conf.c: fix run file for socklog-unix service on solaris; use
    runit's chpst instead of setuidgid.
  * doc/install.html: remove obsolete note about setuidgid replacement.
  * doc/examples.html: adapt.
  * doc/index.html: adapt examples.
  * doc/benefits.html: minor.

socklog 1.1.1
Mon, 07 Jul 2003 09:59:08 +0200
  * socklog-conf.c: fix typos (thx Hleil Liu); use mail -s instead of
    mailsubj for the notify service.

socklog 1.1.0
Thu, 19 Jun 2003 15:40:52 +0200
  * man/socklog-conf.8, man/socklog.8, man/tryto.1: typos.
  * doc/benefits.html, doc/configuration.html, doc/examples.html,
    doc/index.html, doc/network.html, doc/notify.html,
    doc/readme.solaris.html, man/socklog-conf.8, man/socklog.8, man/tryto.1,
    man/uncat.1: utilize runit package instead of daemontools package.
  * socklog-conf.c: create service directories for the use with the runit
    package.

socklog 1.0.0
Thu, 30 Jan 2003 13:32:24 +0100
  * doc/configuration.html, doc/examples.html: remove description for
    socklog-ucspi-unix service.
  * doc/benefits.html, doc/network.html, doc/readme.solaris.html,
    doc/usedietlibc.html, man/socklog-conf.8: typos.
  * error.h: include <errno.h>.
  * tryto.c: catch possible error in write() to child; minor cleanup.
  * doc/index.html: add NetBSD to supported systems.

socklog 0.10.1
Thu, 08 Aug 2002 22:55:35 +0200
  * package/versions: new.

src/nanoklogd.c
/* nanoklogd - forward kernel messages to /dev/log.  */

/* Written by Christian Neukirchen <chneukirchen@gmail.com>
   To the extent possible under law, the creator of this work has waived
   all copyright and related or neighboring rights to this work.  */

/* This uses /dev/log directly because glibc syslog(3) does
   forward LOG_KERN as LOG_USER.  */

#include <sys/klog.h>
#include <sys/socket.h>
#include <errno.h>
#include <unistd.h>

static const struct {
        short sun_family;
        char sun_path[9];
} log_addr = {
        AF_UNIX,
        "/dev/log"
};

int
main(int argc, char *argv[]) {
	char buf[16384];
	int f, l, start;
	int stop = 0;

	f = socket(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0);
	if (f < 0)
		return 111;

	if (connect(f, (void *)&log_addr, sizeof log_addr) < 0)
		return 112;

	while (1) {
		l = klogctl(2, buf, sizeof buf);
		if (l < 0)
			return 113;

		for (start = 0; start < l; start = stop+1) {
			for (stop = start+1; stop < l && buf[stop] != '\n'; stop++)
				;

			while (send(f, buf+start, stop-start, 0) != stop-start) {
				switch (errno) {
				case ECONNREFUSED:
				case ECONNRESET:
				case ENOTCONN:
					/* syslogd went down, retry until up.  */
					close(f);
					sleep(1);
					f = socket(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0);
					if (f < 0)
						return 111;
					connect(f, (void *)&log_addr, sizeof log_addr);
					break;
				default:
					return 114;
				}
			}
		}
	}
	return 0;
}



src/socklog-conf.c
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <pwd.h>
#include <grp.h>
#include <stdio.h>
#include "error.h"
#include "strerr.h"
#include "open.h"
#include "buffer.h"

#if defined(__sun__) && defined(__sparc__) && defined(__unix__) && defined(__svr4__)
#define SOLARIS
#include <sys/utsname.h>
#include "scan.h"
#endif

#ifndef SOLARIS
#include <paths.h>
#endif
#ifndef _PATH_KLOG
#define _PATH_KLOG "/dev/klog"
#endif

#define USAGE1 " unix|inet|klog|ucspi acct logacct [/etc/sv] [/logdir]"
#define USAGE2 " notify acct grp [/etc/sv] [pipe]"

#define VERSION "$Id: socklog-conf.c,v 1.14 2006/03/06 12:56:34 pape Exp $"
#define WARNING "socklog-conf: warning: "
#define FATAL "socklog-conf: fatal: "

#define CONF_DIR "/etc/sv"
#define LOG_DIR_UNIX "/var/log/socklog"
#define LOG_DIR_INET "/var/log/socklog-inet"
#define LOG_DIR_UCSPI_TCP "/var/log/socklog-ucspi-tcp"
#define LOG_DIR_KLOG "/var/log/socklog-klog"
#define PATH_NOTIFY "/var/log/socklog/.notify"

#define CONF_UNIX 0
#define CONF_INET 1
#define CONF_UCSPI_TCP 2
#define CONF_NOTIFY 3
#define CONF_KLOG 4

const char *progname;

void usage() {
  strerr_warn3("usage: ", progname, USAGE1, 0);
  strerr_die4x(1, "usage: ", progname, USAGE2, "\n");
}

const char *dir;
const char *fn;
char buf[1024];
int fd;
buffer b;

char *user;
char *loguser;
const char *path;

struct passwd *upw, *pw;
struct group *gr;

void fail() {
  if (fn[0] == '/') 
    strerr_die4sys(111, FATAL, "unable to create ", fn, ": ");
  else
    strerr_die6sys(111, FATAL, "unable to create ", dir, "/", fn, ": ");
}

void start(const char *s) {
  fn = s;
  fd = open_trunc(fn);
  if (fd == -1) fail();
  buffer_init(&b, buffer_unixwrite, fd, buf, sizeof buf);
}

void outs(const char *s) {
  if (buffer_puts(&b, s) == -1) fail();
}

void finish(void) {
  if (buffer_flush(&b) == -1) fail();
  if (fsync(fd) == -1) fail();
  close(fd);
}

void perm(int mode) { if (chmod(fn, mode) == -1) fail(); }

void mkfile(const char *f, const char *s) {
  start(f);
  outs(s); outs("\n");
  finish();
}

void makedir(const char *s) {
  fn =s;
  if (mkdir(fn, 0750) == -1) fail();
}
void makechdir(const char *s) {
  makedir(s);
  if (chown(s, pw->pw_uid, pw->pw_gid) == -1)
    strerr_die6sys(111, FATAL, "unable to set owner of ", dir, "/", s, ": ");
}

void conf_unix() {
#ifdef SOLARIS
#if WANT_SUN_DOOR
  struct utsname u;
  unsigned long sunos_version;
  uname(&u);
  scan_ulong(u.release+strlen(u.release)-1, &sunos_version);
#endif
#endif
  makedir("socklog-unix");
  perm(0750);
  makedir("socklog-unix/log");
  perm(0755);

  makechdir(path);
  if (symlink(path, "socklog-unix/log/main") == -1)
    strerr_die4sys(111, FATAL, "unable to link ", path, ": ");
  makechdir("socklog-unix/log/main/auth");
  mkfile("socklog-unix/log/main/auth/config", "s999999\nn5\n-*\n+auth.*\n+authpriv.*");
  makechdir("socklog-unix/log/main/cron");
  mkfile("socklog-unix/log/main/cron/config", "s999999\nn5\n-*\n+cron.*");
  makechdir("socklog-unix/log/main/daemon");
  mkfile("socklog-unix/log/main/daemon/config", "s999999\nn5\n-*\n+daemon.*");
  makechdir("socklog-unix/log/main/debug");
  mkfile("socklog-unix/log/main/debug/config", "s999999\nn5\n-*\n+*.debug*");
  makechdir("socklog-unix/log/main/ftp");
  mkfile("socklog-unix/log/main/ftp/config", "s999999\nn5\n-*\n+ftp.*");
  makechdir("socklog-unix/log/main/kern");
  mkfile("socklog-unix/log/main/kern/config", "s999999\nn5\n-*\n+kern.*");
  makechdir("socklog-unix/log/main/local");
  mkfile("socklog-unix/log/main/local/config", "s999999\nn5\n-*\n+local.*");
  makechdir("socklog-unix/log/main/mail");
  mkfile("socklog-unix/log/main/mail/config", "s999999\nn5\n-*\n+mail.*");
  makechdir("socklog-unix/log/main/main");
  mkfile("socklog-unix/log/main/main/config", "s999999\nn10");
  makechdir("socklog-unix/log/main/news");
  mkfile("socklog-unix/log/main/news/config", "s999999\nn5\n-*\n+news.*");
  makechdir("socklog-unix/log/main/syslog");
  mkfile("socklog-unix/log/main/syslog/config", "s999999\nn5\n-*\n+syslog.*");
  makechdir("socklog-unix/log/main/user");
  mkfile("socklog-unix/log/main/user/config", "s999999\nn5\n-*\n+user.*");
  
  start("socklog-unix/run");
  outs("#!/bin/sh\n");
  outs("exec 2>&1\n");
#ifndef SOLARIS
  outs("exec chpst -U"); outs(user); outs(" socklog unix /dev/log\n");
#else
  outs("exec chpst -U"); outs(user); outs(" socklog sun_stream /dev/log");
#if WANT_SUN_DOOR
  if (sunos_version == 7) outs(" /etc/.syslog_door");
  if (sunos_version >= 8) outs(" /var/run/syslog_door");
#endif
  outs("\n");
#endif
  finish();
  perm(0750);

  start("socklog-unix/check");
  outs("#!/bin/sh\n");
  outs("exec 2>/dev/null\n");
  outs("exec socklog-check unix /dev/log\n");
  finish();
  perm(0755);
		  
  start("socklog-unix/log/run");
  outs("#!/bin/sh\n");
  outs("exec chpst -u");
  outs(loguser);
  outs(" svlogd \\\n");
  outs("  main/main main/auth main/cron main/daemon main/debug main/ftp \\\n");
  outs("  main/kern main/local main/mail main/news main/syslog main/user\n");
  finish();
  perm(0750);
}

void conf_inet() {
  makedir("socklog-inet");
  perm(0750);
  makedir("socklog-inet/log");
  perm(0755);

  makechdir(path);
  if (symlink(path, "socklog-inet/log/main") == -1)
    strerr_die4sys(111, FATAL, "unable to link ", path, ": ");
  makechdir("socklog-inet/log/main/main");
  mkfile("socklog-inet/log/main/main/config", "s999999\nn10");

  start("socklog-inet/run");
  outs("#!/bin/sh\n");
  outs("exec 2>&1\n");
  outs("exec chpst -U"); outs(user); outs(" socklog inet 0 514\n");
  finish();
  perm(0750);

  start("socklog-inet/log/run");
  outs("#!/bin/sh\n");
  outs("exec chpst -u");
  outs(loguser);
  outs(" svlogd -t main/main\n");
  finish();
  perm(0750);
}

void conf_ucspi_tcp() {
  makedir("socklog-ucspi-tcp");
  perm(0750);
  makedir("socklog-ucspi-tcp/log");
  perm(0755);

  makechdir(path);
  if (symlink(path, "socklog-ucspi-tcp/log/main") == -1)
    strerr_die4sys(111, FATAL, "unable to link ", path, ": ");
  makechdir("socklog-ucspi-tcp/log/main/main");
  mkfile("socklog-ucspi-tcp/log/main/main/config", "s999999\nn10");

  start("socklog-ucspi-tcp/run");
  outs("#!/bin/sh\n");
  outs("PORT=10116\n");
  outs("exec 2>&1\n");
  outs("exec tcpsvd -vllogserver -u");
  outs(user);
  outs(" 0 \"$PORT\" socklog ucspi TCPREMOTEIP\n");
  finish();
  perm(0750);

  start("socklog-ucspi-tcp/log/run");
  outs("#!/bin/sh\n");
  outs("exec chpst -u");
  outs(loguser);
  outs(" svlogd -t main/main\n");
  finish();
  perm(0750);
}

void conf_klog() {
  makedir("socklog-klog");
  perm(0750);
  makedir("socklog-klog/log");
  perm(0755);

  makechdir(path);
  if (symlink(path, "socklog-klog/log/main") == -1)
    strerr_die4sys(111, FATAL, "unable to link ", path, ": ");
  makechdir("socklog-klog/log/main/main");
  mkfile("socklog-klog/log/main/main/config", "s999999\nn10");

  start("socklog-klog/run");
  outs("#!/bin/sh\n");
  outs("exec <"); outs(_PATH_KLOG); outs("\n");
  outs("exec 2>&1\n");
  outs("exec chpst -u");
  outs(user);
  outs(" socklog ucspi\n");
  finish();
  perm(0750);

  start("socklog-klog/log/run");
  outs("#!/bin/sh\n");
  outs("exec chpst -u");
  outs(loguser);
  outs(" svlogd -t main/main\n");
  finish();
  perm(0750);
}

void conf_notify() {
  makedir("socklog-notify");
  perm(0755);

  umask(007);
  if (mkfifo(path, 0620) == -1)
    strerr_die4sys(111, FATAL, "unable to create \"", path, "\": ");
  umask(022);
  if (chown(path, upw->pw_uid, gr->gr_gid) == -1)
    strerr_die4sys(111, FATAL, "unable to set owner of ", path, ": ");

  start("socklog-notify/run");
  outs("#!/bin/sh -e\n");
  outs("MAILTO=root\n");
  outs("PIPE="); outs(path); outs("\n\n");
  outs("if [ ! -p \"$PIPE\" ]; then mkfifo -m0620 \"$PIPE\"; chown ");
  outs(user); outs(":"); outs(loguser);
  outs(" \"$PIPE\"; fi\n");
  outs("exec <> \"$PIPE\"\n");
  outs("exec chpst -u");
  outs(user);
  outs(" uncat -s49999 -t180 \\\n");
  outs("  env MAILUSER=log MAILNAME='socklog notify' \\\n");
  outs("    mail -s socklog-notify $MAILTO\n");
  finish();
  perm(0750);
}

int main(int argc, char **argv) {
  int mode =0;

  progname =*argv++;
  umask(022);

  if (! *argv) usage();
  switch (**argv) {
  case 'u':
    if (! *(++*argv)) usage();
    switch (**argv) {
    case 'n':
      mode =CONF_UNIX;
      break;
    case 'c':
      mode =CONF_UCSPI_TCP;
      break;
    default:
      usage();
    }
    break;
  case 'i':
    mode =CONF_INET;
    break;
  case 'n':
    mode =CONF_NOTIFY;
    break;
  case 'k':
    mode =CONF_KLOG;
    break;
  case '-':
    if ((*argv)[1] && (*argv)[1] == 'V') {
      strerr_warn2(VERSION, "\n", 0);
    }
  default:
    usage();
  }
  argv++;

  user =*argv++;
  if (!user) usage();
  loguser =*argv++;
  if (!loguser) usage();
  dir =*argv++;
  if (!dir) {
    dir =CONF_DIR;
  } else {
    if (dir[0] != '/') usage();
    if (*argv) {
      path =*argv;
      if (path[0] != '/') usage();
    }
  }

  upw =getpwnam(user);
  if (!upw) strerr_die3x(111, FATAL, "unknown account ", user);
  if (mode != CONF_NOTIFY) {
    pw =getpwnam(loguser);
    if (!pw) strerr_die3x(111, FATAL, "unknown account ", loguser);
  } else {
    gr =getgrnam(loguser);
    if (!gr) strerr_die3x(111, FATAL, "unknown group ", loguser);
  }

  if (chdir(dir) == -1) {
    if (errno != error_noent)
      strerr_die4sys(111, FATAL, "unable to switch to ", dir, ": ");

    if (mkdir(dir, 0700) == -1)
      strerr_die4sys(111, FATAL, "unable to create ", dir, ": ");
    if (chmod(dir, 0750) == -1)
      strerr_die4sys(111, FATAL, "unable to set mode of ", dir, ": ");
    if (chdir(dir) == -1)
      strerr_die4sys(111, FATAL, "unable to switch to ", dir, ": ");
  }
  
  switch(mode) {
  case CONF_UNIX:
    if (!path) path =LOG_DIR_UNIX;
    conf_unix();
    break;
  case CONF_INET:
    if (!path) path =LOG_DIR_INET;
    conf_inet();
    break;
  case CONF_UCSPI_TCP:
    if (!path) path =LOG_DIR_UCSPI_TCP;
    conf_ucspi_tcp();
    break;
  case CONF_KLOG:
    if (!path) path =LOG_DIR_KLOG;
    conf_klog();
    break;
  case CONF_NOTIFY:
    if (!path) path =PATH_NOTIFY;
    conf_notify();
    break;
  }
  return(0);
}
